pkg/codegen/schema/loader.go

```diff
@@ -210,3 +250,106 @@ func LoadPackageReference(loader Loader, pkg string, version *semver.Version) (P
 	}
 	return p.Reference(), nil
 }
+
+func (l *pluginLoader) loadSchemaBytes(pkg string, version *semver.Version) ([]byte, *semver.Version, error) {
+	pluginInfo, err := workspace.GetPluginInfo(workspace.ResourcePlugin, pkg, nil, false /* skipMetadata */)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	if version == nil {
+		version = pluginInfo.Version
+	}
+
+	cachedSchemaPath := filepath.Join(pluginInfo.Path, fmt.Sprintf("schema-%v.json", pkg))
+
+	schemaBytes, ok := l.loadCachedSchemaBytes(pkg, cachedSchemaPath, pluginInfo.InstallTime)
+	if ok {
+		return schemaBytes, nil, nil
+	}
+
+	schemaBytes, provider, err := l.loadPluginSchemaBytes(pkg, version)
+	if err != nil {
+		return nil, nil, fmt.Errorf("Error loading schema from plugin: %w", err)
+	}
+
+	// TODO(friel) disallow caching when the plugin isn't installed in a well known location, i.e.: in the plugin path
+	err = os.WriteFile(cachedSchemaPath, schemaBytes, 0600)
+	if err != nil {
+		return nil, nil, fmt.Errorf("Error writing schema from plugin to cache: %w", err)
+	}
+
+	if version == nil {
+		info, err := provider.GetPluginInfo()
+		if err != nil {
+			// Nonfatal
+		}
+		version = info.Version
+	}
+
+	return schemaBytes, version, nil
+}
+
+func (l *pluginLoader) loadPluginSchemaBytes(pkg string, version *semver.Version) ([]byte, plugin.Provider, error) {
+	if err := l.ensurePlugin(pkg, version); err != nil {
```

Comment:
Although this is pre-existing code, I admit that I'm a little surprised that it works for our tests. I suppose we must use alternative `Loader`s rather than alternative `plugin.Host`s, as there are certainly implementations of the latter that do not use out-of-process plugins (mainly used for by `lifecycletest` for testing the engine). I would have expected this to live in the host implementation.

Reply:
It turns out our integration tests in this repo install 3 different aws plugins, 3 google native plugins, 2 random plugins, a kubernetes plugin, and an azure-native plugin.

Our TestGeneratePackage tests use schemas that refer to external packages. In `GeneratePackageFilesFromSchema` we call `schema.ImportSpec` which doesn't take a loader as an argument, so we rely on the default plugin loader. 

Moving this down here was why my tests were failing (my comment to @iwahbe.) because I was asking the host to resolve the plugin before we had ensured it. My more recent commits move this to right above where we call `l.host.ResolvePlugin` (formerly `workspace.GetPluginInfo`).

😭